File: 0056_merge.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        start, end = intervals[0]
        answer = []
        for i in range(1, len(intervals)):
            if intervals[i][0] > end:
                answer.append([start, end])
                start, end = intervals[i]
            elif intervals[i][0] <= end and intervals[i][1] > end:
                end = intervals[i][1]
            elif intervals[i][0] <= end and intervals[i][1] <= end:
                continue
        answer.append((start, end))
        return answer

# Alternative solution

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key = lambda x: x[0])
        result = [intervals[0]]
        for current in intervals[1:]:
            last = result[-1]
            if last[1] >= current[0]: 
                start = last[0]
                end = max(current[1], last[1])
                result[-1] = [start, end]
            else:
                result.append(current)
        return result

File: 0001-0500/0056_merge.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0], b[0]));
        List<int[]> result = new ArrayList<>();
        result.add(intervals[0]);
        int[] last;
        int[] curr;

        for (int i = 1; i < intervals.length; i++) {
            last = result.get(result.size() - 1);
            curr = intervals[i];

            if (curr[0] <= last[1]) {
                result.set(result.size() - 1, new int[] {last[0], Math.max(last[1], curr[1])});
            } else {
                result.add(curr);
            }
        }

        return result.toArray(new int[result.size()][2]);
    }
}
